### DISCUSSION TITLE
Add an item_hovered() signal to ItemList

### DISCUSSION THREAD

**Discussion by Sithoid** [godotengine/godot-proposals#10412] [GUI]
Preamble: I was trying to change the mouse cursor to "pointing hand" if and only if the item I'm hovering over is "not disabled". For that, I needed an item_hovered signal. Thankfully, there is [a GDScript solution](https://www.reddit.com/r/godot/comments/1egxfwp/how_to_emit_a_signal_when_an_item_in_an_itemlist/lfvh6xl/) on reddit.

However, I feel like:
1) Being able to detect that an item was hovered is a common enough expectation, for various behaviors and use cases;
2) Since "item hovered" styles have been [added to the themes last year](https://github.com/godotengine/godot-proposals/issues/6478), I suspect this is already being detected by the core.

I'm not familiar with C++ so I can't open an issue, but for the reasons above I believe this signal should be exposed by the base class, and it should be a reasonably easy fix.

**Comment by AThousandShips** [comment]
See also:
* https://github.com/godotengine/godot-proposals/issues/10025

**Comment by surreal6** [comment]
Hi. I am working in an accesible UI with text to speech and head selection (gaze pointer for XR interfaces) and found that there is no mouse_entered and mouse_exited signals for optionButton items. There is a item_focused,  so i can implement text to speech when using key input, but i need a item_mouse_entered to detect gaze selection of each item, as it works by moving the mouse over the ui. 

I already have boolean inputs. integer and float sliders, and buttons working with both focus and mouse, godot just lack of this signal to have a complete set of ui interactions.

So, +1 to add  "item_hovered", or "item_mouse_entered" and "item_mouse_exited" signals, This would be a great feature to complete the ui set and to help build accesible interfaces

**Reply by Sithoid** [reply]
So, it turns out that the situation for **[OptionButton]** is even worse because it relies on a Popup, which doesn't extend Control and thus lacks crucial signals. I was able to write hover detection for it based on that example for ItemList, here it goes:

```
# Based on hover detection for ItemList by ForkedStill

extends OptionButton

signal item_hovered(index : int)

# Boundaries of the popup menu
var popup_rect : Rect2
# Height of one menu item
var item_height : int
# Index of the currently hovered item
var hovered_item : int = -1

# Connect popup signals
func _ready() -> void:
	get_popup().mouse_entered.connect(hover_setup)
	get_popup().window_input.connect(hover_detect)
	get_popup().mouse_exited.connect(hover_clear)

# Setup all measurements on mouse enter, assuming they don't change while the popup is open
func hover_setup() -> void:
	print("Mouse entered the popup menu")
	popup_rect = Rect2(get_popup().position, get_popup().size)
	# This only takes item sizes into account. Might need tweaking for separators and margins
	item_height = popup_rect.size.y / get_popup().item_count

# Called when mouse is moving within the list
func hover_detect(event : InputEvent) -> void:
	if event is InputEventMouseMotion:
		var mouse_pos := get_global_mouse_position()
		# This check looks redundant but gives the popup time to avoid division by zero
		if popup_rect.has_point(mouse_pos):
			var index : int = int(mouse_pos.y - popup_rect.position.y) / item_height
			if hovered_item != index:
				if index in range(get_popup().item_count):
					item_hovered.emit(index)
					print("Hovered item %s" % index)
				hovered_item = index

# Called when mouse leaves the list
func hover_clear() -> void:
	print("Mouse left the popup menu")
	hovered_item = -1
```

However, I think we're now properly in "hack" territory. Since hovering over those items is clearly detected (they visually change and also have "Hover" theme properties), I believe that it should be possible (and much needed!) to expose that detection, as well.

Edit: more code fixes

**Reply by surreal6** [reply]
Thanks! 

I found a little error in your code, in the **hover_detect** function, the mouse positions  should be like this:

`var mouse_pos := get_local_mouse_position()

if popup_rect.has_point(get_global_mouse_position()):
....
	`

yours only works if node is in the upper-left corner  :P

**Reply by Sithoid** [reply]
Ah! I knew I should've tested it more. Fixed a bit differently:
```
var mouse_pos := get_global_mouse_position()
if popup_rect.has_point(mouse_pos):
	var index : int = int(mouse_pos.y - popup_rect.position.y) / item_height
```
This also gets rid of the "magic" -1 which seems cleaner.

**Comment by passivestar** [comment]
Needs to be two signals - `item_mouse_entered` and `item_mouse_exited` to be most useful

See also:
- https://github.com/godotengine/godot-proposals/issues/1977

Same is missing from `Tree`, and I noticed wanting to have it more than once. It can be useful when making editor tools for quick previews, and also for things like drawing buttons only on hover to avoid cluttering the UI:

https://github.com/user-attachments/assets/e219824a-4dd1-42f9-b98a-900ed42a2760

The example above uses `Tree.get_item_at_position()`

btw neither `get_item_at_position()` or `has_point()` are enough because it doesn't account for when a tree or itemlist is partially occluded by other controls. But as mentioned above since both trees and itemlists have internal hover detection in engine it probably shouldn't be hard to add `mouse_entered` and `mouse_exited` signals
